streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -149,12 +160,16 @@ public void register(final StateStore store,
                 restoreCallbacks.put(topic, stateRestoreCallback);
                 recordConverters.put(topic, recordConverter);
             } else {
-                log.trace("Restoring state store {} from changelog topic {} at checkpoint {}", storeName, topic, checkpointableOffsets.get(storePartition));
+                final Long restoreCheckpoint = store.persistent() ? initialLoadedCheckpoints.get(storePartition) : null;
```

Comment:
This is where we're using the _loaded_ checkpoint for store registration. Note the missing condition which is now handled... if the store is not persistent, it should _not_ use the loaded checkpoint.